### ISSUE TITLE
App crashing because of high memory usage

### ISSUE THREAD

**Issue by zebganzo** [apollographql/apollo-ios#1662]
## Bug report

Hi there! As a follow up of another [bug](https://github.com/apollographql/apollo-ios/issues/1651) reported, @blizzd and I have gathered some info about the high memory usage when fetching the items. 

## Versions

- apollo-ios SDK version: 0.38.3
- Xcode version: 12.4
- Swift version: Swift 5
- Package manager: CocoaPods

## Further details

Hi Apollo team!

I'm working on the integration of the Apollo client in our app, migrating from a networking layer that was based on "old fashion" POST requests. I'm facing some issues because the app is crashing because of the very **high usage of the memory**.

.Context.

Our app has to fetch from the server a few thousand elements. For several reasons, we had to split requests in bulks of 300. Thus, if the app needs to download 1500 items, it will create 5 separated requests that, afterwards, will be combined together with their output.

.Gist of the technical implementations.

It's based on ReactiveSwift. Hopefully, even with little familiarity with it, this should be digestible by anyone.

```
let fetchingSignalProducers: Array<SignalProducer<[Item], NoError>> = ...

SignalProducer<[Item], RequestError>
                .combineLatest(fetchingSignalProducers)
                .map { arrayOfarrayOfItems in
                	...
                }
```

Each one of the fetchingSignalProducers performs a fetch operation, running on a dedicated queue. The code above is running on a background, dedicated queue.

.Investigation.

We have been digging for a few days, using the `Debug navigator`, the `Debug Memory Graph` and the instruments.

These screenshots are from a run of the app stopped at a quite low level of usage compared to those that get to ~2GB.

![Screenshot 2021-02-16 at 14 51 49](https://user-images.githubusercontent.com/3023626/108091647-34431800-707c-11eb-87b7-ab070dcc2c97.png)

![Screenshot 2021-02-16 at 14 52 06](https://user-images.githubusercontent.com/3023626/108091995-897f2980-707c-11eb-8be2-4b1e62726897.png)

![Screenshot 2021-02-16 at 14 51 31](https://user-images.githubusercontent.com/3023626/108092040-956aeb80-707c-11eb-8dcf-0117e263435c.png)

![Screenshot 2021-02-16 at 14 52 18](https://user-images.githubusercontent.com/3023626/108092074-9f8cea00-707c-11eb-9243-189fccc57446.png)

<img width="914" alt="Screenshot 2021-02-16 at 15 44 16" src="https://user-images.githubusercontent.com/3023626/108092101-a4ea3480-707c-11eb-8cd5-68460c2a038b.png">

<img width="1416" alt="Screenshot 2021-02-16 at 15 44 54" src="https://user-images.githubusercontent.com/3023626/108092107-a7e52500-707c-11eb-9bae-89c43291207d.png">

From our investigation, it seems like the issue is whit this Array of PossiblyDeferred that creates a reference cycle.

Which other info can we provide for helping to fix this behaviour?

**Comment by blizzd** [comment]
Here are some more precise stacks.
![PossiblyDeferred_swift2](https://user-images.githubusercontent.com/2085575/108093421-ffd05b80-707d-11eb-93f5-a704a0009a66.png)
![PossiblyDeferred_swift1](https://user-images.githubusercontent.com/2085575/108093438-02cb4c00-707e-11eb-93f5-54dc9d690a9a.png)
![GraphQLExecutor_swift](https://user-images.githubusercontent.com/2085575/108093451-06f76980-707e-11eb-82a8-084e7ec8f4b7.png)

**Comment by designatednerd** [comment]
Hm, that stack trace for a deferred operation definitely seems odd, since theoretically the thunk shouldn't be recursively calling `get`, and definitely something VERY goofy going on if there's 121k `Node` objects, which theoretically should be quickly gotten rid of. 

We'll look into it, thank you!

**Comment by designatednerd** [comment]
One thing @martijnwalraven pointed out about the `PossiblyDeferred` structure is that every single field on an object will return a `PossiblyDeferred`, which is probably why you're seeing so many calls to `get()` there. 

In regard to the many many `Node` instances, there definitely may be a lot - as @martijnwalraven put it to me: 

>  every level of a response would add another Node, and having lists of nested objects with lists, etc. adds up.

And it does sound like you have a fair number of nested objects in a list of like 1500 things. In theory those should get released after use though, so i'd be curious to see if they're not. 

Other questions: 
- Where are you putting the breakpoint that's showing these stacktraces? 
- Why does the app need to fetch all 1500 elements at once rather than lazy loading them? 
- Are you using in-memory storage or SQLite storage? 
- Is there any possibility you could share your project with me confidentially, or at least set up a repro project?

**Comment by zebganzo** [comment]
Hi @designatednerd!

It is indeed nested, here is the `graphql` [file](https://github.com/apollographql/apollo-ios/files/5995531/Sale.graphql.zip).

Let me try to reply to your questions.

- there was no specific breakpoint, we pressed pause once noticed the memory leak and then dug into the threads
- the app needs to work offline, so we have to synch the local database with the remote one. This ~1500 elements synch happens only at the login/signup when the user doesn't have anything cached
- we are using in-memory at the point of the memory leak, only after the whole procedure succeded the app then stores them with Realm
- we can't share the project because it has too many sensitive info. We could set up a fake project with the same structure, but it could take a few days.

Is there anything else we can try out for providing you with more info? Is having a call an option?

**Comment by blizzd** [comment]
Hi @designatednerd !

I hope to add some valuable insight in this. I may have found the reason why the memory does not get free.
This also made me dive a bit deeper : 
![RetainCycle-GraphQLSelection](https://user-images.githubusercontent.com/2085575/108235074-c5ca8c80-7145-11eb-894e-e5583bb1f050.png)

Now, it looks like your `Zip3Accumulator` which is using `GraphQLResultAccumulator`
`GraphQLResultAccumulator -> func zip<Accumulator1: GraphQLResultAccumulator, Accumulator2: GraphQLResultAccumulator, Accumulator3: GraphQLResultAccumulator>(_ accumulator1: Accumulator1, _ accumulator2: Accumulator2, _ accumulator3: Accumulator3) -> Zip3Accumulator<Accumulator1, Accumulator2, Accumulator3>`
is using a protocol with an `associatedtype PartialResult` for it, which the function attempts to verify and accept results, but at the same time is allocating memory for each next result, while holding the previous result in memory. O( N x N ).

https://stackoverflow.com/questions/44765347/does-enum-retain-its-associated-object

P.S. Experimenting with SQLStorage instead of In-Memory Storage only _delayed_ the above effect.

**Comment by designatednerd** [comment]
Associated **objects** (which are what's being used with enums, and what that SO link is about) and associated **types** (which are generally used with generic protocols, and what we're dealing with here) are two different things. 

However, we'll definitely take a deeper look at the memory implications of these zip functions, particularly because I think there may be some `zip` functions we can take advantage of in Swift that didn't exist when these were written.

**Comment by designatednerd** [comment]
> We could set up a fake project with the same structure, but it could take a few days.

Honestly, this would probably be the most helpful option because we can then see _exactly_ what's going on. A call is an option but it sounds like this is buried pretty deep into things. 

> the app needs to work offline, so we have to synch the local database with the remote one. This ~1500 elements synch happens only at the login/signup when the user doesn't have anything cached

Have you tried to flush to Realm after each 300-item parse is done rather than waiting for all 1500 items to show up?

**Comment by zebganzo** [comment]
@designatednerd where can I send you the project? I have created a fake one that only fetches the items. No (cool) UI,  no Realm, etc.

The app went up to 2.28GB of meemory.

**Comment by designatednerd** [comment]
ellen at apollographql dot com. Thank you!

**Comment by zebganzo** [comment]
Thanks Ellen!

I set it up in a way that it has only the bare minimum logic from my side and then the Apollo client. It has reproduced the same behaviours and memory usage and computational stress.

![Screenshot 2021-02-19 at 10 26 24](https://user-images.githubusercontent.com/3023626/108501131-1c080e80-72b1-11eb-8587-f8d582175849.png)
![Screenshot 2021-02-19 at 10 26 35](https://user-images.githubusercontent.com/3023626/108501142-20ccc280-72b1-11eb-9352-702b763271c0.png)

I also tried to isolate as much as possible **a single request** instead of 5 or 6 in sequence. Here below you can see in the first one some reactive logic, instead the second one has literally only the Apollo request.

![Only1](https://user-images.githubusercontent.com/3023626/108501350-7608d400-72b1-11eb-9278-0060cb689b10.png)
![Only1_NoReactive](https://user-images.githubusercontent.com/3023626/108501358-799c5b00-72b1-11eb-9ab1-399c10f585f5.png)

I find a bit strange both the spike in the memory usage but also the memory left allocated afterwards. I hope my debug prints around will help you out.

This issue is a critical blocker for us because such a high memory usage makes the app crashes when run on a physical device.

About the project I'm going to send you, here are the main classes and their duties:

```
ApolloNetworkWrapper // Apollo initialization and fetch method
SalesRepository // Initilises the ApolloNetworkWrapper and performs the mapping to the models
UpdateSalesService // It implements the logic for the fetch of the items
ViewController // It simply triggers the fetch, either with or without the business logic.
```

**Comment by designatednerd** [comment]
Great, thank you, I got your email and I'll look at this Monday!

**Comment by designatednerd** [comment]
When profiling using the `self.noReactiveTest` I was able to confirm that while it goes up to about 400 megs of memory usage, nothing is actually _leaking_, everything eventually gets deallocated.

I did notice when trying to examine the response in [Charles](https://www.charlesproxy.com/), Charles absolutely choked when trying to load the data into the `JSON Text` view to the point where I had to force quit it. I've never seen that before. I did a bit more examination and after several other apps freezing on me, I found formatting the over 3kb of JSON from that, it's over _71 thousand_ lines of JSON for 300 items. 

Just for giggles I tried switching to batches of 30 instead of 300 and...it went a lot faster and only went up to 170 megs of memory:

<img width="866" alt="Screen Shot 2021-02-22 at 10 12 47 PM" src="https://user-images.githubusercontent.com/1976498/108801396-39600580-755b-11eb-8601-dbb873fc0d77.png">

Can you confirm you're seeing the same behavior on your end if you switch from 300 to 30 items per batch?

**Comment by zebganzo** [comment]
Hi @designatednerd, yes, the behaviour is better with smaller batches. With our current sales set, that means that ~55 requests though.

I'm still wondering how come that out of 300 items, it gets to use 400MB. Isn't it happening for other companies that are using the library?
Also, the amount of MB kept used in the memory after the fetch is still pretty high. How can we free up that?

**Comment by designatednerd** [comment]
@zebganzo I just filed #1683 which outlines a bunch of the problems underlying this performance issue. I think for now I'd prefer to track that as the long-term issue. 

If it's possible, we'd love to grab some kind of test server from your team to let us keep using that project for testing - it's a huge help for us to have a real-world data source that can return gargantuan amounts of data like this.

**Comment by Elaz-viz** [comment]
@zebganzo I can confirm the exact same thing happens to me as well. After days of debugging and profiling, digging into the memory graph got me to the exact same snapshot. Did you end up with a workaround for it?

**Comment by zebganzo** [comment]
Hi @Elaz-viz, I sadly removed the Apollo SDK from the project. At the current state, it is a solution that doesn't fit in our scenario.

**Comment by Elaz-viz** [comment]
@zebganzo thanks for your reply - i hope you find a good solution.
@designatednerd I actually just found a more accurate cause for this memory bloat; the case is where two requests of the same type are being processed by the LegacyParser, on two different threads. There's no thread safety installed on the interceptors, so this causes some kind of race condition in parsing that just keeps creating objects. Having a complicated graph with dependencies surely helps, but I can say 100% that once I protected the request in terms of concurrency, the problem disappeared, so I advise you look in this direction...

**Comment by AnthonyMDev** [comment]
Thanks for that info @Elaz-viz! We'll look into this!

**Comment by MrGobert** [comment]
Getting similar issues as OP running Profiler. My app never seems to release the memory so it can quickly ballon to over 2GB of memory used. We're starting to see more and more of our users crash for Out of Memory errors because of this.

**Comment by Tilfah2** [comment]
Hello, I am facing a similar issue, the memory is usage is increasing to 2-3 GB, but is not being released after and causing the app to crash.

**Comment by AnthonyMDev** [comment]
If either of you can confirm what @Elaz-viz said with your setups, that would be really helpful. Try putting your operations in a serial queue and see if that makes it go away.

I have a suspicion that there are two separate issues going on here, but not sure what's causing this still.

The 1.0 version we are working on rewrites a large part of the execution layer, so I'm wondering if these problems will just go away with that version.

**Comment by Tilfah2** [comment]
Putting operations in a serial queue did not solve the problem. The only conclusion that I reached is the memory rising proportional to the number of object stored in normalized cache.

**Comment by rjuzaszek** [comment]
Hello, any updates on this, it just happened to me as well (I realized it happens actually when started to download larger data batches), which together with freezing the UI issue is making a feature not usable for my company :(?

**Comment by AnthonyMDev** [comment]
It sounds like there are multiple different issues occurring here. There are a number of different possible causes of high memory usage, not all of them solvable in the same way. 

**If you can confirm that the high memory usage is related to the `Node` object indicated in this bug report, it is more likely related.** Otherwise, it may be an entirely different issue.

If you are able to migrate to try using the [1.0 Alpha Release](https://github.com/apollographql/apollo-ios/releases/tag/1.0.0-alpha.4) that we are testing and confirming that this is still an issue there, it would be really helpful to us. We hope the 1.0 migration isn't too much of a heavy lift for everyone, but it will need to be done eventually once it's a stable release anyways.

> The only conclusion that I reached is the memory rising proportional to the number of object stored in normalized cache.
@Tilfah2 Are you using an in-memory normalized cache, or a SQL cache? We may need to do more consideration of limiting the size of the in-memory cache, but that's going to be a relatively large project.

> freezing the UI issue is making a feature not usable for my company 
@rjuzaszek Your execution should never be done on the main thread, so I'm not sure why it's freezing the UI. Can you provide more information?

**Comment by Tilfah2** [comment]
@AnthonyMDev I am using in-memory cache normalized cache
